### ISSUE TITLE
Introduce automated notification of new version availability to admin of GlobaLeaks

### ISSUE THREAD

**Issue by fpietrosanti** [globaleaks/GlobaLeaks#1493]
This ticket is to introduce automated notification of new version availability to admin of GlobaLeaks, so that admins can be notified when a new version of the software is released.

GlobaLeaks should poll a URL somewhere managed by us, that contain the new version information and message that the software shall fetch, parse and send by email to the admin of the platform.

That way it will be possible to notify all the admin upon the releases, concentrating the public / unmanaged upgraded within a specific timeframe window.

**Comment by NSkelsey** [comment]
If you are looking to generate an email to the admin of the platform, why not just send them one directly from an external account. . . . . . this leaks a bit of info and certainly doesn't handle admin's changing hands well, but it is certainly the easiest way to handle it.

Can the platform itself receive emails? Also a TOR HS polling some static IP sounds like it could lead to deanonymization issues.

**Comment by fpietrosanti** [comment]
Well, but doing as specified we don't know the "admin emails" but we'll be just providing public information on which is the new version and the new message to be shared along with the Changelog.

It will be entirely up to the GlobaLeaks software to fetch that public information (that yes, can be also done over Tor, if available), know the admin email, know his email server, and notify the admin.

That way we'll not collect any kind of personal/private information about the globaleaks users

**Comment by fpietrosanti** [comment]
We need to consider releasing the software using the Github Release procedure, then having GlobaLeaks fetching new software versions availability from Github servers using Github Release APIs https://developer.github.com/v3/repos/releases/ .

**Comment by fpietrosanti** [comment]
Interesting that when using Github releases all the information are available as RSS atom, providing a platform for integration into many distribution channel for the informations on software releases:
https://www.reddit.com/r/3dshacks/comments/4bg436/tutorial_get_email_notifications_for_new_releases/
https://github.com/blog/1547-release-your-software
Interesting "conventions" on Release Changelog https://github.com/olivierlacan/keep-a-changelog/issues/1#issuecomment-54014136

**Comment by evilaliv3** [comment]
totally up: to use standard guidelinees for Changelogs like keep-a-changelog (as suggested in past by @mmaker i'm already following various grammar/format/uniformity rules).

doubts: this will increase an overall quality but wont wont wont automate anything; the most important failing component in making it possible to automate a changelog is THE DEVELOPER. if the developer does not author send out clean commits with an explicit, standard, uniformed title and more important split commits including only what the title of the commit does, the automated changelog will be just shit.
to cope with this issues, relevant softwares (e.g. angularjs), make the development process pass forcefully from PULL REQUEST and follows strict rules of naming while editing each commit before integration in the master branch. doing this (that is not at all an automated process) the result is great: https://github.com/angular/angular/blob/master/CHANGELOG.md

**Comment by fpietrosanti** [comment]
@evilaliv3 afaik writing the changelog should still be done manually, also in angularjs they refer to pull request title, but all the structure and intermediary description is done manually. 

The uses of github releases IMHO gives the advantage of keeping all of the code, binaries and release information into the github platform in the github way including release information, changelog, source-code tarbal and released binaries available from github. 

Then the convention/template of changelog is up to us, if referenced to pull or to issues, if more tech or more user-oriented. I saw a kind of changelog structure nice on https://github.com/spf13/hugo/releases . We could make up a template with all the sections that *must* be filled while making a release, including writing "NONE" if there's no change of a specific section?

**Comment by evilaliv3** [comment]
Agreed.

What remains to me not clear is why archiving here the deb packages when
anyhow we will require an external debian repository.
At the same time it is not clear to me the order of the release; Probably
if the announcement is done via Github and GlobaLeaks makes the
notification using the github ATOM we will have to perform the GitHub
release at last step in our release procedure.

**Comment by fpietrosanti** [comment]
Idea: This ticket could be implemented by downloading and parsing the file "Packages" from the latest updated Debian repository, using existing python library to parse the Release versions.

**Comment by evilaliv3** [comment]
@fpietrosanti the software in python for parsing the repository files is: python-debian

it is in xenial but it is very very simply implemented with not so good functionalities;
e.g. applied to our repository it just read the first package it finds in the file and giving it's information

will be on us so to implement to spit that file parsing each of the record and determining the latest version available because it does not offer for example such a functionality.

**Comment by evilaliv3** [comment]
@NSkelsey: this would be actually a good ticket where to implement:
- fetch of the repository RELEASE file
- parse of the last version available
- notification to the admin (marking when the notification happened so to be able to repeat it only after a certain period); for this we could already prepare a text and put it under translations.

**Comment by fpietrosanti** [comment]
For a quick release, we can also provide only visibility of the latests release version in the main admin interface, making the notification a secondary ticket (along with it's configurability to enable/disable the notification)

**Comment by NSkelsey** [comment]
I agree with @fpietrosanti. We can add visibility that the application can be updated and handle separately the notification when we choose to be more aggressive about telling users to update.

In response to @evilaliv3 idea to download the `RELEASE` file to detect the latest version is an okay solution. The [python-debian](https://packages.debian.org/sid/python-debian) package will __always__ be supported for ubuntu and debian, but hardcoding in the hostname deb.globaleaks.org and adding parsing of debian repo files manually are not great.

If we just had some way to ensure that `sudo apt-get update` was also run periodically this would be a solved issue. We could just schedule a job to run a one line bash command.
```bash
> apt-cache policy globaleaks | awk '/Candidate/ {print $2}'
```

As an aside, a nice benefit of the work that is going to go into either implementation is that we can take the version detection and port it into the positions where we detect if the application has been updated in `models.config.update_defaults`.

**Comment by NSkelsey** [comment]
With the package `unattended-upgrades` installed as a dependency and it's configuration aligned to only issue updates we can trade less python for more packaging complexity.

The file `/etc/apt/apt.conf.d/20auto-upgrades` needs to be changed only slightly. Further, I think when issuing the install these two configuration variables can be automatically specified.
```
APT::Periodic::Update-Package-Lists "1";
APT::Periodic::Unattended-Upgrade "0";
```

**Comment by fpietrosanti** [comment]
Example uses https://github.com/romlok/python-debian/blob/master/README.deb822

**Comment by NSkelsey** [comment]
Current (1/8/17) version of [python-debian](https://packages.ubuntu.com/xenial/python-debian) in xenial is v1.27-2.

**Comment by NSkelsey** [comment]
Here is what the UI will look like for administrators. We do have the date from the latest release because the Packages file does not include such information. It would be nice to have, but we can do without it.

![version-notification](https://user-images.githubusercontent.com/1095794/28831156-2014c1ee-76d9-11e7-9d90-11add7c36987.png)

I was tossing around the idea of including the change-log of the latest missing release here as well, but it would not help very much because their could be many updates between the current installed version and the current latest update.

**Comment by NSkelsey** [comment]
I would like for feedback on 2723fec. 

I believe it addresses this ticket adequately with the possibility of extending it to notifying the administrator via mail. It needs review before release that @evilaliv3 has offered to do.

**Comment by evilaliv3** [comment]
Going to pick up the code review and doing it by tonight.

**Comment by NSkelsey** [comment]
This work has been reviewed in 0ef63c323d and will be released with `v2.70.11`

**Comment by fpietrosanti** [comment]
being a security feature, should also be documented in document https://docs.google.com/document/d/1SMSiAry7x5XY9nY8GAejJD75NWg7bp7M1PwXSiwy62U/pub

**Comment by evilaliv3** [comment]
@NSkelsey / @fpietrosanti: Im re-opening this ticket in order to keep track of and improvement inline with the first definition of this ticket.

Currently the presence of a new release is communicated to the admin only by means of the web interface.

This ticket is to extend such a communication to happen via Email as for other kind of events.

While implementing this feature we should evaluate how to keep track of events that have been triggered new version presence and mail notification and conditions in which skip/re-trigger the system action (e..g. the mail notification.

**Comment by evilaliv3** [comment]
I've written the template for the notifications that are now under translation.

We could now proceed by adding a variable "last_version" were to store the last version read from the repository; the system should, in case would identify that a new version is available, notify the user that a new version is available. The mail should be sent to every administrator of the platform.

**Comment by NSkelsey** [comment]
New code is up which adds the update mail logic in fd094e796080bba0b3733b1ad463da36f65b79f5 Only one mail will get sent per user per software update to keep the number of mails the machine is sending out low.

I'll wait for a review before closing ticket anew.

**Comment by fpietrosanti** [comment]
Is this complete or there's still some todo? @evilaliv3

**Comment by fpietrosanti** [comment]
Hey, that's implemented
